content/kubeone/master/guides/operating-system-manager/_index.en.md

```diff
@@ -0,0 +1,50 @@
+.
+This has been the norm in KubeOne till KubeOne v1.4.x and it works as expected. Although over the time, it has been observed that this workflow has certain limitations.
+
+## Machine Controller Limitations
+
+- Machine Controller expects **ALL** the supported user-data plugins to exist and be ready. User might only be interested in a subset of the available operating systems. For example, user might only want to work with `ubuntu`.
+- The user-data plugins have templates defined [in-code](https://github.com/kubermatic/machine-controller/blob/master/pkg/userdata/ubuntu/provider.go#L133). Which is not ideal since code changes are required to update those templates. Then those code changes need to become a part of the subsequent releases for machine-controller and KubeOne. So we need a complete release cycle to ship those changes to customers.
+- Managing configs for multiple cloud providers, OS flavors and OS versions, adds a lot of complexity and redundancy in machine-controller.
+- Since the templates are defined in-code, there is no way for an end user to customize them to suit their use-cases.
+- Each cloud provider sets some sort of limits for the size of `user-data`, machine won't be created in case of non-compliance. For example, at the time of writing this, AWS has set a [hard limit of 16KB](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-add-user-data.html).
+- Better support for air-gapped environments is required.
+
+Operating System Manager was created to overcome these limitations.
+
+## Architecture
+
+OSM introduces the following new resources:
+
+### OperatingSystemProfile
+
+Templatized resource that represents the details of each operating system. OSPs are immutable and default OSPs for supported operating systems are provided/installed automatically by kubeOne. End users can create custom OSPs as well to fit their own use-cases.
+
+### OperatingSystemConfig
+
+Immutable resource that contains the **actual configurations** that are going to be used to bootstrap and provision the worker nodes. It is a subset of OperatingSystemProfile, OperatingSystemProfile is a template while OperatingSystemConfig is an instance rendered using OperatingSystemProfile, MachineDeployment and flags provided at OSM level.
+
+OperatingSystemConfigs have a 1-to-1 relation with the MachineDeployment. A dedicated controller watches the MachineDeployments and generates the OSCs in `kube-system` and secrets in `cloud-init-settings` namespaces in the cluster. Machine Controller then waits for the bootstrapping and provisioning secrets to become available. Once they are ready, it will extract the configurations from those secrets and pass them as `user-data`
+to the machines that it's going to provision.
```

Comment:
```suggestion
OperatingSystemConfigs have a 1-to-1 relation with the MachineDeployment. A dedicated controller watches the MachineDeployments and generates the OSCs in `kube-system` and secrets in `cloud-init-settings` namespaces in the cluster. Machine Controller then waits for the bootstrapping- and provisioning-secrets to become available. Once they are ready, it will extract the configurations from those secrets and pass them as `user-data` to the to-be-provisioned machines.
```